{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "5013a0f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot Guide Counts and Generate Fold Change Table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0434ec59",
   "metadata": {},
   "outputs": [],
   "source": [
    "# start coding here\n",
    "import os\n",
    "import math\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from sklearn.metrics import auc\n",
    "\n",
    "%matplotlib inline\n",
    "\n",
    "sns.set_style('ticks')\n",
    "sns.set_context('talk')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "110fe625",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Read in raw read counts per sample\n",
    "\n",
    "Generated by `combine_read_counts.py` and `read_counts_from_bam.py` scripts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2f82685d",
   "metadata": {},
   "outputs": [],
   "source": [
    "read_counts = pd.read_csv(snakemake.input[0])\n",
    "samp2mapped = read_counts.groupby(['sample_name'])['Coverage'].sum().to_dict()\n",
    "read_counts['fraction_reads'] = read_counts.apply(\n",
    "    lambda x: x.Coverage/samp2mapped[x.sample_name],\n",
    "    axis=1)\n",
    "read_counts.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9db863de",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Determine T0 Samples from sample sheet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4ab740e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "read_counts.sample_name.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ae332034",
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_df = pd.read_csv(snakemake.config['samples'], sep='\\t')\n",
    "sample_df.sort_values(by=['timepoint',], inplace=True)\n",
    "sample_order = sample_df.sample_name.values\n",
    "sample_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c5143e3e",
   "metadata": {},
   "outputs": [],
   "source": [
    "t0 = sample_df.timepoint.min()\n",
    "t0_samples = sample_df[sample_df.timepoint == t0].sample_name.values\n",
    "t0_samples"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "41a57ebd",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Plot Lorenz Curve for T0 Samples\n",
    "\n",
    "Closer to 0.5 the better"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b4a1af20",
   "metadata": {},
   "outputs": [],
   "source": [
    "pal = sns.color_palette('colorblind', len(t0_samples))\n",
    "fig, axes = plt.subplots(figsize=(13, 5), ncols=len(t0_samples))\n",
    "for smpl, ax in zip(t0_samples, axes):\n",
    "    y = read_counts[read_counts.sample_name == smpl].fraction_reads.sort_values(\n",
    "        ascending=False).values\n",
    "    x = np.arange(0, len(y)) / len(y)\n",
    "    y = np.cumsum(y)\n",
    "    ax.plot(x, y, c=pal.pop(0))\n",
    "    ax.plot([0, x[-1]], [0, y[-1]], '--', c='grey')\n",
    "    lbl = auc(x, y)\n",
    "    ax.text(0, 0.9, \"AUC = {:.3f}\".format(lbl))\n",
    "    ax.set_title(smpl)\n",
    "    ax.set_xlabel(\"Ranked Guides\")\n",
    "    if smpl == t0_samples[0]:\n",
    "        ax.set_ylabel(\"Fraction Reads\")\n",
    "plt.savefig(snakemake.output[0])\n",
    "prefix, _ = os.path.splitext(snakemake.output[0])\n",
    "plt.savefig(prefix + '.png')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "096ab72e",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Plot Read Distributions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "86efe1eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "g = sns.FacetGrid(read_counts,\n",
    "                  row='sample_name',\n",
    "                  row_order=t0_samples,\n",
    "                  hue='sample_name',\n",
    "                  hue_order=t0_samples,\n",
    "                  sharex=False,\n",
    "                  sharey=False,\n",
    "                  height=6,\n",
    "                  aspect=1.5)\n",
    "g.map_dataframe(sns.histplot, 'Coverage')\n",
    "plt.savefig(snakemake.output[1])\n",
    "prefix, _ = os.path.splitext(snakemake.output[1])\n",
    "plt.savefig(prefix + '.png')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "582b8bfe",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "g = sns.FacetGrid(read_counts,\n",
    "                  row='sample_name',\n",
    "                  row_order=sample_order,\n",
    "                  hue='sample_name',\n",
    "                  hue_order=sample_order,\n",
    "                  sharex=False,\n",
    "                  sharey=False,\n",
    "                  height=6,\n",
    "                  aspect=1.5)\n",
    "g.map_dataframe(sns.histplot, 'Coverage')\n",
    "plt.savefig(snakemake.output[2])\n",
    "prefix, _ = os.path.splitext(snakemake.output[2])\n",
    "plt.savefig(prefix + '.png')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "79f687b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Calculate Normalized Read Counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d8b1100f",
   "metadata": {},
   "outputs": [],
   "source": [
    "read_counts['normalized_count'] = read_counts.fraction_reads * 1e7\n",
    "read_counts.normalized_count.replace(0, 0.1, inplace=True)\n",
    "pivot = read_counts.pivot_table(index='Target',\n",
    "                                columns='sample_name'\n",
    "                               )['normalized_count'].reset_index()\n",
    "pivot.index.name = 'index'\n",
    "pivot['Gene'] = pivot.Target.apply(lambda x: x.split('_')[3])\n",
    "col_order = ['Target', 'Gene'] + list(sample_order)\n",
    "pivot = pivot[col_order]\n",
    "pivot.to_csv(snakemake.output[3], sep='\\t', index=False)\n",
    "pivot.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "07eacf6d",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Calculate Fold Change vs T0 for each Genotype\n",
    "\n",
    "The resulting tab-delimited file can be used with BAGEL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b9832b9b",
   "metadata": {},
   "outputs": [],
   "source": [
    "fold_change_df = pivot[['Target', 'Gene']].copy()\n",
    "for gt in sample_df.genotype.unique():\n",
    "    gt_df = sample_df[sample_df.genotype == gt]\n",
    "    smp_t0 = gt_df[gt_df.timepoint == t0].sample_name.values\n",
    "    if len(smp_t0) != 1:\n",
    "        raise ValueError(\"Expected exactly 1 T0 sample for genotype {}\".format(gt))\n",
    "    comp_samples = gt_df[gt_df.sample_name != smp_t0[0]].sample_name.values\n",
    "    for smpl in comp_samples:\n",
    "        print(\"Comparing {} vs {}\".format(smpl, smp_t0[0]))\n",
    "        fold_change_df[smpl] = (pivot[smpl]/pivot[smp_t0[0]]).apply(lambda x: math.log2(x))\n",
    "fold_change_df.to_csv(snakemake.output[4], sep='\\t', index=False)\n",
    "fold_change_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fcd12042",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "7b77513020dd5b1722f5f1a5228e302cb01f97f3bf519056f9c70bb06bff16b5"
  },
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
